flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -673,18 +601,32 @@ private void setChaining() {
                     info.getStartNodeId(),
                     1, // operators start at position 1 because 0 is for chained source inputs
                     info,
-                    chainEntryPoints);
+                    chainEntryPoints,
+                    true,
+                    serializationExecutor,
+                    jobVertexBuildContext,
+                    null);
         }
     }
 
-    private List<StreamEdge> createChain(
+    public static List<StreamEdge> createChain(
             final Integer currentNodeId,
             final int chainIndex,
             final OperatorChainInfo chainInfo,
-            final Map<Integer, OperatorChainInfo> chainEntryPoints) {
+            final Map<Integer, OperatorChainInfo> chainEntryPoints,
+            final boolean canCreateNewChain,
+            final Executor serializationExecutor,
+            final JobVertexBuildContext jobVertexBuildContext,
+            final @Nullable Consumer<Integer> hashGenerator) {
 
         Integer startNodeId = chainInfo.getStartNodeId();
-        if (!builtVertices.contains(startNodeId)) {
+        if (!jobVertexBuildContext.getJobVerticesInOrder().containsKey(startNodeId)) {
+            StreamGraph streamGraph = jobVertexBuildContext.getStreamGraph();
+
+            // Progressive hash generation is required in adaptive graph generator.
+            if (hashGenerator != null) {
+                hashGenerator.accept(currentNodeId);
```

Comment:
From the perspective of this method, it is not a generator because it generates nothing.